packages/python/m/github/actions/action_file.py

```diff
@@ -0,0 +1,144 @@
+from m.core import yaml
+from m.pydantic import KebabModel
+from pydantic import BaseModel
+from pydantic.fields import PydanticUndefined
+
+from .actions import Action
+from .misc import MetadataOutput
+
+TEMPLATE = """\
+# AUTOGENERATED FILE - Do not update by hand!
+# Edit {py_path}/{module_name}.py
+#   then run `m github actions [python_file]` to update
+
+name: {name}
+description: {description}
+
+inputs:{inputs}
+outputs:{outputs}
+runs:
+  using: composite
+  steps:
+    {steps}
+"""
+EMPTY_DICT = ' {}'  # noqa: P103 - not a format string
+
+
+class TemplateVars(BaseModel):
+    """Template variables."""
+
+    py_path: str
+
+    module_name: str
+
+    name: str
+
+    description: str
+
+    inputs: str
+
+    outputs: str
+
+    steps: str
+
+
+def _indent(text: str, num: int) -> str:
+    spaces = '  ' * num
+    return text.replace('\n', f'\n{spaces}', -1)
+
+
+class ActionFile(BaseModel):
+    """Helper class to write the action file."""
+
+    py_path: str
+
+    module_name: str
+
+    name: str
+
+    description: str
+
+    action: Action
+
+    inputs: type[KebabModel]
+
+    outputs: dict[str, MetadataOutput]
+
+    python_path: str
+
+    available_outputs: dict[str, str]
+
+    def build_inputs(self: 'ActionFile') -> str:
+        """Generate the inputs for the action.
+
+        Returns:
+            A string to add to the Github action.
+        """
+        all_inputs: dict[str, dict[str, str | bool]] = {}
```

Comment:
```suggestion
        all_inputs: input_class = {}
```

that's a complicated enough type signature that it might be a good idea to make an interface to keep track of it, just to make it easier to keep track of

Reply:
named it `InputInfo`. A dictionary was used as an aggreator to dump all the values, it is only an implementation detail but it is worth having that hint of what we are doing here.